### PR TITLE
Restructure repository

### PR DESCRIPTION
**Pull request by qaristote**
This PR implements the feature suggested in #19.

### CODE REVIEW

**Comment by facundominguez** [comment]
Ah, so before the modules in `smtlib-backends-tests` where compiled twice. Once in `smtlib-backends-tests` and once in `smtlib-backends`.

Perhaps dependencies could be arranged as follows.

`smtlib-backends` is at the bottom and doesn't depend on anything else.
`smtlib-backends-tests` depends on `smtlib-backends`
`smtlib-backends-z3` depends on `smtlib-backends` and `smtlib-backends-tests`
A new package `smtlib-backends-process` depends on `smtlib-backends` and `smtlib-backends-tests` :)

I think it is possible to reexport modules in `smtlib-backends-process` if you don't want users to list both `smtlib-backends-process` and `smtlib-backends` in their dependencies.

**Comment by qaristote** [comment]
> I think it is possible to reexport modules in smtlib-backends-process if you don't want users to list both smtlib-backends-process and smtlib-backends in their dependencies.

I did that but that creates a cycle `smtlib-backends <-> smtlib-backends-process`, which is probably harder to deal with than the `smtlib-backends.test-suite -> smtlib-backends-tests -> smtlib-backends`.

```
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: smtlib-backends-0.1 (user goal)
[__1] trying: smtlib-backends-process-0.1 (user goal)
[__2] rejecting: smtlib-backends-process:*test (cyclic dependencies; conflict
set: smtlib-backends, smtlib-backends-process)
[__1] fail (backjumping, conflict set: smtlib-backends,
smtlib-backends-process)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: smtlib-backends,
smtlib-backends-process
```

**Comment by facundominguez** [comment]
I just contributed a commit to break the circular dependency.

File: smtlib-backends.cabal [hunk 1]
```diff
@@ -28,32 +28,10 @@ library
   ghc-options:      -Wall -Wunused-packages
   exposed-modules:
     SMTLIB.Backends
-    SMTLIB.Backends.Process
```
Comment (qaristote) [inline smtlib-backends.cabal:31]: But now the user has to add both `smtlib-backends` and `smtlib-backends-process` as dependencies, don't they ?

Comment (facundominguez) [reply smtlib-backends.cabal:31]: Correct. 

You also could use [this cabal feature](https://cabal.readthedocs.io/en/3.4/cabal-package.html#pkg-field-library-reexported-modules) to expose `SMTLIB.Backends` in `smtlib-backends-process`. But I'm not sure if it is worth the trouble.

Comment (qaristote) [reply smtlib-backends.cabal:31]: I tried adding
```
reexported-modules: smtlib-backends-process:SMTLIB.Backends.Process
```
in `smtlib-backends.cabal` but somehow I get
```
Resolving dependencies...
Error:
    Problem with module re-exports:
      - The package 'smtlib-backends-process'
        does not export a module 'SMTLIB.Backends.Process'
    In the stanza 'library'
    In the inplace package 'smtlib-backends-0.1'
```

So unless I made an obvious mistake, yeah it's probably not worth the trouble ...

Comment (facundominguez) [reply smtlib-backends.cabal:31]: You are trying to expose a module of `smtlib-backends-process` in `smtlib-backends`. I think `reexpoted-modules` allow you to do the converse: exposing `SMTLIB.Backends` in `smtlib-backends-process`.

**Pull request merged**